be/src/storage/lake/gc.cpp

```diff
@@ -13,120 +13,155 @@
+    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(root_location));
+    auto txn_log_root_location = join_path(root_location, kTxnLogDirectoryName);
+    auto iter_st = fs->iterate_dir(txn_log_root_location, [&](std::string_view name) {
+        if (is_txn_log(name)) {
+            auto [tablet_id, txn_id] = parse_txn_log_filename(name);
+            if (txn_id < min_active_txn_id && owned_tablets.count(tablet_id) > 0) {
+                VLOG(2) << "Deleting " << name;
+                auto location = join_path(txn_log_root_location, name);
+                auto st = fs->delete_file(location);
+                LOG_IF(WARNING, !st.ok() && !st.is_not_found()) << "Fail to delete " << name << ": " << st;
             }
-            return true;
-        });
-        if (iter_st.is_not_found()) {
-            // ignore this error
-            return Status::OK();
-        }
-
-        if (!iter_st.ok()) {
-            return iter_st;
         }
+        return true;
+    });
+    return iter_st.is_not_found() ? Status::OK() : iter_st;
+}
 
-        for (const auto& filename : txn_logs) {
-            auto location = join_path(txn_log_root_location, filename);
-            auto st = fs->delete_file(location);
-            LOG_IF(WARNING, !st.ok() && !st.is_not_found()) << "Fail to delete " << location << ": " << st;
-        }
+static Status drop_disk_cache(std::string_view root_location) {
+    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(root_location));
+    auto segment_root_location = join_path(root_location, kSegmentDirectoryName);
+    auto orphan_list_location = join_path(root_location, kOrphanListFileName);
+    auto file_or = fs->new_random_access_file(orphan_list_location);
+    if (file_or.status().is_not_found()) {
+        return Status::OK();
+    } else if (!file_or.ok()) {
+        return file_or.status();
+    }
+    std::vector<std::string> orphan_segments;
+    RETURN_IF_ERROR(read_orphan_list_file(file_or->get(), &orphan_segments));
+    for (const auto& seg : orphan_segments) {
+        VLOG(3) << "Dropping disk cache of " << seg;
+        auto location = join_path(segment_root_location, seg);
+        // TODO: Add a new interface for dropping disk cache
+        auto st = fs->delete_file(location);
+        LOG_IF(WARNING, !st.ok() && !st.is_not_found()) << "Fail to drop cache of " << seg << ": " << st;
     }
-
     return Status::OK();
 }
 
-Status segment_gc(std::string_view root_location, TabletManager* tablet_mgr) {
-    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(root_location));
+Status metadata_gc(std::string_view root_location, TabletManager* tablet_mgr, int64_t min_active_txn_id) {
+    const auto owned_tablets = tablet_mgr->owned_tablets();
+    Status ret;
+    ret.update(delete_tablet_metadata(root_location, owned_tablets));
+    ret.update(delete_txn_log(root_location, owned_tablets, min_active_txn_id));
+    ret.update(drop_disk_cache(root_location));
```

Comment:
only one BE can gc metadata and segment, but all BE should drop disk cache?

Reply:
Yes